package/MDAnalysis/lib/picklable_file_io.py

```diff
@@ -0,0 +1,190 @@
+
+class FileIOPicklable(io.FileIO):
+    """File object (read-only) that can be pickled.
+
+    This class provides a file-like object (as returned by :func:`open`, namely :class:`io.FileIO`)
+    that, unlike standard Python file objects, can be pickled. Only read mode is supported.
+
+    When the file is pickled, filename and position of the open file handle in the file
+    are saved. On unpickling, the file is opened by filename and the file is
+    seeked to the saved position. This means that for a successful unpickle, the original
+    file still has to be accessible with its filename.
+
+    Example
+    -------
+    ::
+        file = FileIOPicklable('filename')
+        file.readline()
+        file_pickled = pickle.loads(pickle.dumps(file)
+        assert_equal(file.tell(), file_pickled.tell())
+
+    See Also
+    ---------
+    TextIOPicklable
+    BufferIOPicklable
+    """
+    def __init__(self, name):
+        super().__init__(name, mode='r')
+
+    def __getstate__(self):
+        return self.name, self.tell()
+
+    def __setstate__(self, args):
+        name = args[0]
+        super().__init__(name, mode='r')
+        self.seek(args[1])
+
+
+class BufferIOPicklable(io.BufferedReader):
+    """A picklable buffer object for read-only FilIO object.
+
+    This class provides a buffered :class:`io.BufferedReader` object that can be pickled. Note that this only works in reda mode.
+
+    Example
+    -------
+    ::
+        file = FileIOPicklable('filename')
+        buffer_wrapped = BufferIOPicklable(file)
+
+    See Also
+    ---------
+    FileIOPicklable
+    TextIOPicklable
+    """
+    def __init__(self, raw):
+        super().__init__(raw)
+        self.raw_class = raw.__class__
+
+    def __getstate__(self):
+        return self.raw_class, self.name, self.tell()
+
+    def __setstate__(self, args):
+        raw_class = args[0]
+        name = args[1]
+        raw = raw_class(name)
+        super().__init__(raw)
+        self.seek(args[2])
+
+
+class TextIOPicklable(io.TextIOWrapper):
+    """Character and line based picklable file-like object.
+
+    This class provides a file-like :class:`io.TextIOWrapper` object that can
+    be pickled. Note that this only works in read mode.
+
+    Example
+    -------
+    ::
+        file = FileIOPicklable('filename')
+        text_wrapped = TextIOPicklable(file)
+
+    See Also
+    ---------
+    FileIOPicklable
+    BufferIOPicklable
+    """
+    def __init__(self, raw):
+        super().__init__(raw)
+        self.raw_class = raw.__class__
+
+    def __getstate__(self):
+        return self.raw_class, self.name, self.tell()
+
+    def __setstate__(self, args):
+        raw_class = args[0]
+        name = args[1]
+        # raw_class is used for further expansion this functionality to
+        # GZip files, which also requires a text wrapper.
+        raw = raw_class(name)
+        super().__init__(raw)
+        self.seek(args[2])
+
+
+def pickle_open(name, mode='rt'):
```

Comment:
I do not particularly like this function. It is unclear from the name that this can only be used to open files for reading. You also implicitly assume we always want to pickle a file object when we open it for reading. I would much rather have an exception *when* we actually try to pickle the file.

Reply:
So if we get IO classes that behave like the original ones except that they raise an error when you try to pickle with “w” but pickle fine with “r” then we still need our own `open()` function, right? Could we directly have `anyopen()` use these new classes? (Maybe you said this already somewhere...)

Reply:
Not sure. But yeah only using anyopen uniformly in MDAnalysis sounds good.

Reply:
We could mark the `pickle_open()` function and related functions as private for now, i.e., `_pickle_open()`, `_bz2_pickle_open()`, `_gsd_pickle_open()` etc. We then have the freedom to change implementation in a backward-incompatible fashion even after a 2.0 release, e.g., if we want to go back to the idea to having classes that accept r and w and only fail pickling at runtime.

Or are ultimately ok with the design as it stands, @kain88-de ?